File: 46.Permutations/Permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
          vector<int> solution;
          vector<vector<int> > result;
          helper(nums, solution, result);
          return result;
    }

    void helper(vector<int> nums, vector<int> solution, vector<vector<int> > &result){
      if(nums.size() == 0){
        result.push_back(solution);
        return;
      }
      for(int i=0;i<nums.size();i++){
        int tmp = nums[i];
        nums.erase(nums.begin() + i);
        solution.push_back(tmp);
        helper(nums, solution, result);
        solution.pop_back();
        nums.insert(nums.begin() + i, tmp);
      }
    }
};

File: 46.Permutations/Permutations.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def helper(nums, solution, result):
            if not nums:
                result.append(solution)
                return
            for index, num in enumerate(nums):
                helper(nums[:index]+nums[index+1:], solution + [num], result)
        solution = []
        result = []
        helper(nums, solution,result)
        return result

s = Solution()
print s.permute([1,1,2])
